<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Проект-по-SQL-—-сервис-для-чтения-книг-по-подписке" data-toc-modified-id="Проект-по-SQL-—-сервис-для-чтения-книг-по-подписке-1">Проект по SQL — сервис для чтения книг по подписке</a></span><ul class="toc-item"><li><span><a href="#Подключение-к-БД-и-получение-данных" data-toc-modified-id="Подключение-к-БД-и-получение-данных-1.1">Подключение к БД и получение данных</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-1.2">Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-1.2.1">Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-1.2.2">Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-1.2.3">Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-1.2.4">Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-1.2.5">Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-1.2.6">Общий вывод</a></span></li></ul></li></ul></li></ul></div>

# Проект по SQL — сервис для чтения книг по подписке

Задача — проанализировать базу данных крупного сервиса для чтения книг по подписке для формулирования ценностного предложения для нового продукта. В базе данных представлена информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

Необходимо будет решить следующие задачи:
- посчитать, сколько книг вышло после 1 января 2000 года;
- для каждой книги посчитать количество обзоров и среднюю оценку;
- определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- определить автора с самой высокой средней оценкой книг;
- посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

*В работе необходимо использовать только язык запросов SQL. Pandas допускается только для хранения и вывода результатов.*

## Подключение к БД и получение данных

In [1]:
# импорт библиотек:
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt  

# устанавка параметров:
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохранение коннектора:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
a = 'SELECT * FROM books'
b = 'SELECT * FROM authors'
c = 'SELECT * FROM publishers'
d = 'SELECT * FROM ratings'
e = 'SELECT * FROM reviews'

# Выведем все таблицы через цикл:
list = [a, b, c, d, e]
for table in list:
    display(pd.io.sql.read_sql(table, con = engine))

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Таблица **books** содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица **authors** cодержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица **publishers** cодержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;


Таблица **ratings** cодержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица **reviews** cодержит данные о пользовательских обзорах:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

## Анализ данных

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [3]:
query = '''

SELECT COUNT(DISTINCT(book_id))       
FROM books
WHERE publication_date > '2000-01-01'

'''
pd.io.sql.read_sql(query, con = engine) 

count
0    819

**Вывод**:<br>
Видим, что после 1 января 2000 года было опубликовано **819 книг.**

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [4]:
query = '''

WITH data AS (

SELECT rat.book_id,
       b.title, 
       COUNT(DISTINCT(review_id)) AS count_reviews,
       AVG(rating) AS avg_rating     
FROM reviews as rev
JOIN ratings as rat ON rat.book_id = rev.book_id
JOIN books as b ON b.book_id = rat.book_id 
GROUP BY rat.book_id,
         b.title
)

SELECT b.book_id,
       d.title,
       d.count_reviews,
       d.avg_rating
from data AS d
RIGHT OUTER JOIN books as b ON b.book_id = d.book_id 


'''
pd.io.sql.read_sql(query, con = engine) 

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      672                                               None   
996       83                                               None   
997      221                                               None   
998      387                                               None   
999      808                                               None   

     count_reviews  avg_rating  
0              2.0    3.666667  
1              1.0    2.500000  
2              3.0    4.666667  
3              2.0    4.500000  
4              4.0    4.000000  
..             ...         ...  
995            NaN         NaN  
996            NaN         NaN  
997            NaN         NaN  
998            NaN         NaN  
999            NaN         NaN  

[1000 rows x 4 columns]

**Вывод**:<br>
Мы вывели информацию о количестве обзоров и средней оценке для каждой книги.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [5]:
query = '''

SELECT b.publisher_id,
       p.publisher
FROM books AS b
JOIN publishers AS p ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY b.publisher_id,
         p.publisher
ORDER BY COUNT(DISTINCT(book_id)) DESC
LIMIT 1

'''
pd.io.sql.read_sql(query, con = engine) 

publisher_id      publisher
0           212  Penguin Books

**Вывод**:<br>
Мы определили издательство, которое выпустило наибольшее число книг толще 50 страниц — **Penguin Books.**

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [9]:
query = '''

SELECT author,
       AVG(rating)
FROM ratings AS r
JOIN books AS b ON b.book_id = r.book_id
JOIN authors AS a ON a.author_id = b.author_id

WHERE b.author_id in (
    SELECT b.author_id
    FROM ratings AS r
    JOIN books AS b ON b.book_id = r.book_id
    GROUP BY b.book_id
    HAVING COUNT(DISTINCT(rating_id)) > 50
)  

GROUP BY b.author_id,
         author
ORDER BY AVG(rating) DESC
LIMIT 1


'''
pd.io.sql.read_sql(query, con = engine) 

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

**Вывод**:<br>
Мы определили автора с самой высокой средней оценкой книг — **J.K. Rowling/Mary GrandPré (средняя оценка составляет 4,29).**

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [7]:
query = '''

WITH data AS (
    SELECT username,
           COUNT(DISTINCT(review_id))
    FROM reviews       
    WHERE username in (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT(book_id)) > 50)
    GROUP BY username) 
SELECT AVG(count)
FROM data

'''
pd.io.sql.read_sql(query, con = engine) 

avg
0  24.333333

**Вывод**:<br>
Мы посчитали среднее количество обзоров от пользователей, которые поставили больше 50 оценок — оно составляет **24,33.**

### Общий вывод

**1.** Цель проекта: Проанализировать базу данных крупного сервиса для чтения книг по подписке и ответить на ряд вопросов.

**2.** В процессе решения поставленых задач мы выяснили следующее:
- после 1 января 2000 года вышло **819 книг**;
- мы вывели для каждой книги количество обзоров и средний рейтинг;
- издательство **Penguin Books** выпустило наибольшее число книг. Мы учитывали только книги толще 50 страниц, чтобы  исключить из анализа брошюры;
- **J.K. Rowling/Mary GrandPré** является автором с самой высокой средней оценкой книг — **4,29**. Мы учитывали только книги с 50 и более оценками;
- В среднем количество обзоров от пользователей, которые поставили больше 50 оценок, составляет **24,33**.